lib/AST/UnqualifiedLookup.cpp

```diff
@@ -492,11 +482,29 @@ ValueDecl *UnqualifiedLookupFactory::ResultFinderForTypeContext::lookupBaseDecl(
   // In these cases, using the Swift 6 lookup behavior doesn't affect
   // how the body is type-checked, so it can be used in Swift 5 mode
   // without breaking source compatibility for non-escaping closures.
-  if (!factory->Ctx.LangOpts.isSwiftVersionAtLeast(6) &&
-      !implicitSelfReferenceIsUnwrapped(selfDecl, closureExpr)) {
+  if (capturesSelfWeakly && !factory->Ctx.LangOpts.isSwiftVersionAtLeast(6) &&
+      !implicitSelfReferenceIsUnwrapped(selfDecl)) {
     return nullptr;
   }
 
+  // Closures are only allowed to rebind self in specific circumstances:
+  //  1. In a capture list with an explicit capture.
+  //  2. In a `guard let self = self` / `if let self = self` condition
+  //     in a closure that captures self weakly.
+  //
+  // These rebindings can be done by any parent closure, and apply
+  // to all nested closures. We only need to check these structural
+  // requirements loosely here -- more extensive validation happens
+  // later in MiscDiagnostics::diagnoseImplicitSelfUseInClosure.
+  //
+  // Other types of rebindings, like an arbitrary "let `self` = foo",
+  // are never allowed to rebind self.
+  if (auto var = dyn_cast<VarDecl>(selfDecl)) {
+    if (!(var->isCaptureList() || var->getParentPatternStmt())) {
+      return nullptr;
```

Comment:
This is the core of the change. To allow an unwrapped `guard let self else { return }` binding to be used as implicit self, the closure has to use the `ASTScope::lookupSingleLocalDecl` lookup behavior for implicit self. Previously we only did this for closures with a `[weak self]` capture, even though any closure could need this behavior.

The majority of the rest of the changes in the PR are a rework required to reject invalid code like https://github.com/apple/swift/pull/70575#issuecomment-1951685335, which was previously rejected thanks to this problematic behavior here where the closures didn't use the newer lookup behavior.